students/tammyd/lesson05/mailroom_pt3.py

```diff
@@ -0,0 +1,171 @@
+            "mandy moore": ("Mandy Moore", [50.99, 100.99, 3000.99]),
+            "jessica simpson": ("Jessica Simpson", [50.99, 100.99, 4000.99]),
+            "queen bey": ("Queen Bey", [50.99, 100.99, 5000.99]),
+            }
+
+def donor_database():
+    return donor_db
+
+
+def mainloop():
+    print("Welcome to the Mailroom \n")
+    main_menu()
+
+
+def main_menu():
+    print("Main Menu:")
+    print(
+      "T - Send a Thank You\n"
+      "L - Send a Letter to all Donors\n"
+      "R - Create a Report\n"
+      "Q - Quit the Program"
+      )
+
+    menu_dict = {"T": thank_you, 
+                 "L": letters_to_everyone, 
+                 "R": create_report, 
+                 "Q": quit_program
+                 }
+
+    while True:
+        menu_selection = input("\nChoose a letter:  ")
+        menu_selection = menu_selection.upper().strip()
+        try:
+            #can you remind me how the () here works?
+            return menu_dict[menu_selection]()
+
+            # if menu_selection == "T":
+            #     thank_you()
+            # elif menu_selection == "L":
+            #     letters_to_everyone()
+            # elif menu_selection == "R":
+            #     create_report()
+            # elif menu_selection == "Q":
+            #     quit_program()
+
+        #why is my exception not being raised here when I make an invalid entry?
+        except (KeyError):
+            print("Invalid response -- select T, R, or Q \n")
+            return 
+        #else:
+            #pass
+        #finally:
+            #pass
+       
+
+def thank_you():
+    print("\nMenu: Send a Thank You\n")
+    thank_you_dict = {"list": donor_database,
+                      "menu": main_menu,
+                      "q": quit_program
+                      }
+
+    while True:
+        donor_name = input("Enter the donor's full name. \n-- or 'list' to see a list of donors, \n-- or 'menu' to return to the main menu: ").strip()
+        donor_name = donor_name.lower()
+        try: 
+            return thank_you_dict[donor_name]()
+
+        # if donor_name == "list":
+        #     print("\nDonor database:")
+        #     for keys in donor_db:
+        #         print(keys.title())
+        #     print("\n")
+        # elif donor_name == "menu":
+        #     return main_menu()
+        # elif donor_name == "q":
+        #     return quit_program()
+        # else:
+        #     break
+        except KeyError:
+            print("Incorrect entry. Try again.")
+            return
+    donation_amount(donor_name)
+
+
+def donation_amount(donor_name):
+    while True:
+        donor_amount = input("Enter amount of donation amount (or 'menu' to exit): ").strip()
+        if donor_amount.lower() == "menu":
+            return main_menu()
+            #what is the difference between these three?...
+               #return
+               #return main_menu
+               #return main_menu()
```

Comment:
On its own, "return" returns nothing, which effectively brings you back to the main menu and that's what you should do to prevent building up the call stack. return main_menu() builds the call stack by calling main menu when you don't need to. and return main_menu returns a function, main menu, without calling it, so that's not what you want.